### PR TITLE
Add Freckle-specific prelude

### PR DESCRIPTION
**Pull request by mjgpy3**
**Why?**

In removing `classy-prelude` from our monorepo, a ton of import ergonomics have
been lost. This attempts to recover some of those ergonomics and also be the
starting point for our very own prelude.

This was mostly inspired by the things I had to import a gazillion types when
removing `classy-prelude` from a few of our libraries. Before hitting the
library that has the most instances, I thought it'd be nice to have something
like this in place.

**Things to look out for**

 - Did I export too much?
 - Are there glaring omissions?
 - Did I pick the wrong thing in some case (e.g. is `Control.Monad.Reader` the
 - "right" reader?)
 - Should this move into our monorepo?
 - Should I get a dep from another package? All new packages were hidden deps, which is nice, but should I have instead just squashed RIO + Prelude?
  - Do _YOU_ specifically have a prelude preference that we should consider instead?
  - Should this be in freckle-app or elsewhere?

**Note on reviewing**

This is likely controversial so I'll be sure to eyes from all interested
backenders.

### CODE REVIEW

**Comment by stackptr** [comment]
Going to follow the conversation on this PR but removing my review to defer to those with more opinions about the prelude

File: library/Freckle/App/Database.hs
Comment (pbrisbin) [inline library/Freckle/App/Database.hs:26]: When doing something like this (using a safe prelude and an unsafe function from `Prelude`). My convention is to `import Prelude as Unsafe (read)`. Then it's nice and clear that you're using `Unsafe.read` later. I also don't consider the `Unsafe` import a "prelude", so I group it down with the rest.

File: library/Freckle/App/Prelude.hs
Comment (pbrisbin) [inline library/Freckle/App/Prelude.hs:82]: If you've got this, let's add `(***)` too?

**Review by pbrisbin — APPROVED** [review]
I don't feel strongly what exactly this looks like; I think we all just wanted to start conservative and expand in response to concrete need; so this LGTM!

I just wonder: just how close did you land to re-creating `Relude` or `RIO.Prelude` here?

File: library/Freckle/App/Database.hs
Comment (mjgpy3) [reply library/Freckle/App/Database.hs:26]: I like that -- it makes things pretty clear.

**Comment by mjgpy3** [comment]
> I just wonder: just how close did you land to re-creating Relude or RIO.Prelude here?

That's a good question, a preliminary glance reveals that this is pretty close.

File: library/Freckle/App/Prelude.hs
Comment (eborden) [inline library/Freckle/App/Prelude.hs:95]: I love that you are hiding some partial functions. Can we export their friends from `Safe` (`headMay`, `readMay`, etc) and `semigroupoids` (`fold1`, `foldMap1`, etc).

File: library/Freckle/App/Prelude.hs
Comment (eborden) [inline library/Freckle/App/Prelude.hs:11]:
```suggestion
  , HashMap
  , HashSet
```

File: library/Freckle/App/Prelude.hs
Comment (eborden) [inline library/Freckle/App/Prelude.hs:49]: Honestly, I'd just export all of `Foldable` and `Traversable`, maybe with the `_1` functions hidden, in favor of their `semigroupoid` version.

File: library/Freckle/App/Prelude.hs
Comment (eborden) [inline library/Freckle/App/Prelude.hs:133]: We have a bunch of modules in `megarepo` like `FrontRow.Core.Applicative` with functions like this that are probably worth migrating here. I don't want that work to block this, but it could be a good followup task.

**Review by eborden — APPROVED** [review]
This looks really solid and matches what we already do, just removing a lot of the pain. I have nothing blocking, just a few thoughts.

We probably want to consider coming up with rules for what gets included in a prelude. Those decision points will arise and we should make it easier on ourselves by having a rubric (maybe with a comment action when this file is changed.)

File: library/Freckle/App/Prelude.hs
Comment (mjgpy3) [reply library/Freckle/App/Prelude.hs:49]: Sounds good but aren't the `_{n}` functions from `Control.Lens`?

**Review by StevenXL — APPROVED** [review]
Really appreciate that the Freckle Prelude has evolved from concrete problems you ran into. LGTM!

File: library/Freckle/App/Prelude.hs
Comment (eborden) [reply library/Freckle/App/Prelude.hs:52]: Oh sorry, I meant things like `foldl1`, not `_1` and `_2`. Maybe `____1` would have been more understandable?

Comment (mjgpy3) [reply library/Freckle/App/Prelude.hs:52]: Oh, gotcha!

**Comment by mjgpy3** [comment]
I think I'm going to remove `on`. Looking into our code a little more we hide it constantly in favor of `Database.Esqueleto (on)`.

**Pull request merged**
